### DISCUSSION TITLE
How to programmatically subscribe a user to a page discussion?

### DISCUSSION THREAD

**Discussion by enzedonline** [danirus/django-comments-xtd#325] [Q&A] [answered]
Hi,

I've tried tracing the process of submitting a comment with the follow-up checkbox ticked to see what the subscribe mechanism is, but don't seem to be able to trap it anywhere.

Is there a method to subscribe a user to a discussion? I'd like to subscribe the page author to the discussion by default so that they are notified when comments are posted on their page. 

Thanks.

**Comment by danirus** [comment, marked as answer]
Hi @enzedonline, it has no such capability. Comments are posted to objects, but the connection between the Comment model and the object receiving the comments is based on the ContentType, it is not enough to know whether the object receiving the comments has a user attribute. When adopting the comments app for your site you can resolve that. I think the quickest solution would be to connect the signal comment_was_posted to a function to notify your users.

**Reply by enzedonline** [reply]
Ahhh. Thanks for that. I was thinking there would be something like a many-to-many table (a bit like the comment_flag table) pairing user_id's with a comment tree root node id. Couldn't see where this was being stored as I would have just added the author ID to that table.

I looked into the signal, but didn't see how I could add my own code into that without editing the xtd code in the virtual environment. In the end, I have a solution in the save method of my custom comment model which stores the parent page id (Wagtail site). I hacked the notify_comment_followers code to send to the author if the author isn't making the comment with an extra context variable to say if it's an author to amend the email text.

```
class CustomComment(XtdComment):
    page = ParentalKey(BlogDetailPage, on_delete=models.CASCADE, related_name='customcomments')

    def save(self, *args, **kwargs):
        if self.user:
            self.user_name = self.user.display_name
        self.page = BlogDetailPage.objects.get(pk=self.object_pk)
        self.notify_author()
        super(CustomComment, self).save(*args, **kwargs)

    def notify_author(self):
        try:
            author = CustomUser.objects.get(id=self.page.owner_id)
        except CustomUser.DoesNotExist:
            return
        if not self.user_email == author.email:
            followers = {}
            followers[author.email] = (
                self.user_name,
                signed.dumps(
                    self.comment, 
                    compress=True,
                    extra_key=settings.COMMENTS_XTD_SALT
                )
            )

            subject = _("New comment on your blog post")
            text_message_template = loader.get_template(
                "django_comments_xtd/email_followup_comment.txt")
            if settings.COMMENTS_XTD_SEND_HTML_EMAIL:
                html_message_template = loader.get_template(
                    "django_comments_xtd/email_followup_comment.html")

            for email, (name, key) in six.iteritems(followers):
                mute_url = reverse('comments-xtd-mute', args=[key.decode('utf-8')])
                message_context = {
                    'user_name': name,
                    'comment': self,
                    'mute_url': mute_url,
                    'site': self.site,
                    'author': True
                }
                text_message = text_message_template.render(message_context)
                if settings.COMMENTS_XTD_SEND_HTML_EMAIL:
                    html_message = html_message_template.render(message_context)
                else:
                    html_message = None
                send_mail(subject, text_message, settings.COMMENTS_XTD_FROM_EMAIL,
                        [email, ], html=html_message)
```
Could be annoying for the author if two commenters go off on a long conversation together, maybe with a bit more development it could be enhanced to allow authors to silence notifications from pages or comment threads. Well, the comment section on these pages is meant as a Q&A forum, so it'll work for now :)

Incidentally, in case it's useful for a future update, while looking at the notify_comment_followers, I noticed that the followers loop overwrites the same key value for each occurrence of every email value. In my test page, there are 3 followers with 60 or so comments between them, so the loop goes 60 times to produce 3 dictionary entries. If you have a compatible database (like POSTGRES), this can be cut down to once per email address using distinct:
```
previous_comments = XtdComment.objects\
			  .filter(**kwargs)\
			  .exclude(user_email=comment.user_email)\
			  .distinct('user_email').order_by('user_email')

for instance in previous_comments:
    followers[instance.user_email] = (....

```
On my page, this cuts the loop from 60 to 3 iterations. Distinct isn't available on all databases (like SQLlite), could be trapped with try: distinct except: not distinct.

Thanks again for your help and quick reply.

**Reply by danirus** [reply]
Good point about "distinct".
Indeed it can be fixed with a try-except.
I will include the change if nobody does it.
Thank you!

**Reply by danirus** [reply]
@enzedonline, your fix on distinct has been merged already in master and released as 2.9.3. Thank you!

**Reply by enzedonline** [reply]
Great! Glad it was useful.
